File: net.jgp.books.spark.ch02/src/main/python/lab100_csv_to_db/csvToRelationalDatabaseApp.py
'''
  CSV to a relational database.

  @author rambabu.posa
'''

from pyspark.sql import SparkSession
from pyspark.sql import functions as F
import os

current_dir = os.path.dirname(__file__)
relative_path = "../../../../data/authors.csv"
absolute_file_path = os.path.join(current_dir, relative_path)

# Creates a session on a local master
spark = SparkSession.builder.appName("CSV to DB").master("local").getOrCreate()

#  Step 1: Ingestion
#  ---------
#
#  Reads a CSV file with header, called authors.csv, stores it in a dataframe
df = spark.read.csv(header=True, inferSchema=True, path=absolute_file_path)

# Step 2: Transform
# ---------
# Creates a new column called "name" as the concatenation of lname, a
# virtual column containing ", " and the fname column
df = df.withColumn("name", F.concat(F.col("lname"), F.lit(", "), F.col("fname")))

# Step 3: Save
# ----
#
# The connection URL, assuming your PostgreSQL instance runs locally on the
# default port, and the database we use is "spark_labs"
dbConnectionUrl = "jdbc:postgresql://localhost/spark_labs"

# Properties to connect to the database, the JDBC driver is part of our pom.xml
prop = {"driver":"org.postgresql.Driver", "user":"jgp", "password":"Spark<3Java"}

# Write in a table called ch02
df.write.jdbc(mode='overwrite', url=dbConnectionUrl, table="ch02", properties=prop)

# Good to stop SparkSession at the end of the application
spark.stop()

File: net.jgp.books.spark.ch02/src/main/java/net/jgp/books/spark/ch02/lab100_csv_to_db/CsvToRelationalDatabaseApp.java
package net.jgp.books.spark.ch02.lab100_csv_to_db;

import static org.apache.spark.sql.functions.concat;
import static org.apache.spark.sql.functions.lit;

import java.util.Properties;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;

/**
 * CSV to a relational database.
 * 
 */
public class CsvToRelationalDatabaseApp {

  public static void main(String[] args) {
    CsvToRelationalDatabaseApp app = new CsvToRelationalDatabaseApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder().appName("CSV to DB").master("local").getOrCreate();

    // Step 1: Ingestion:
    // Reads a CSV file with header, called authors.csv, stores it in a dataframe.
    Dataset<Row> df = spark.read().format("csv").option("header", "true").load("data/authors.csv");

    // Step 2: Transform:
    // Creates a new column called "name" as the concatenation of lname, a virtual
    // column containing ", " and the fname column.
    df = df.withColumn("name", concat(df.col("lname"), lit(", "), df.col("fname")));

    // Step 3: Save:
    // The connection URL, assuming your PostgreSQL instance runs locally on the
    // default port, and the database we use is "spark_labs".
    String dbConnectionUrl = "jdbc:mysql://localhost:3306/spark_labs";

    // Properties to connect to the database, the JDBC driver is part of our
    // pom.xml.
    Properties prop = new Properties();
    prop.setProperty("driver", "com.mysql.cj.jdbc.Driver");
    prop.setProperty("user", "root");
    prop.setProperty("password", "84E4sC2G9g!");

    // Write in a table called ch02.
    df.write().mode(SaveMode.Overwrite).jdbc(dbConnectionUrl, "ch02", prop);

    System.out.println("Process complete");
  }
}
